pkg/scheduler/framework/plugins/clusteraffinity/types.go

```diff
@@ -6,101 +6,298 @@ Licensed under the MIT license.
+			if q.Cmp(expectedQ) <= 0 {
+				// The observed value is less than or equal to the expected one (expected to be
+				// greater than the value).
+				return false, nil
+			}
+		case placementv1beta1.PropertySelectorGreaterThanOrEqualTo:
+			if q.Cmp(expectedQ) < 0 {
+				// The observed value is less than the expected one (expected to be greater
+				// than or equal to the value).
+				return false, nil
+			}
+		case placementv1beta1.PropertySelectorLessThan:
+			if q.Cmp(expectedQ) >= 0 {
+				// The observed value is greater than or equal to the expected one (expected to be
+				// less than the value).
+				return false, nil
+			}
+		case placementv1beta1.PropertySelectorLessThanOrEqualTo:
+			if q.Cmp(expectedQ) > 0 {
+				// The observed value is greater than the expected one (expected to be less than
+				// or equal to the value).
+				return false, nil
+			}
+		default:
+			// The operator is not recognized; normally this should never happen.
+			return false, fmt.Errorf("invalid operator: %s", exp.Operator)
 		}
 	}
-	return score
+	// The cluster matches the property selector.
+	return true, nil
 }
 
-func newAffinityTerm(term *placementv1beta1.ClusterSelectorTerm) (*affinityTerm, error) {
-	selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
+// clusterPreference is a type alias for PreferredClusterSelector in the API, which allows
+// easy method extension.
+type clusterPreference placementv1beta1.PreferredClusterSelector
+
+// interpolateWeightFor interpolates weight based on the observed value of a property.
+func interpolateWeightFor(cluster *clusterv1beta1.MemberCluster, property string, sortOrder placementv1beta1.PropertySortOrder, weight int32, state *pluginState) (int32, error) {
+	q, err := retrievePropertyValueFrom(cluster, property)
 	if err != nil {
-		return nil, err
+		return 0, fmt.Errorf("failed to perform weight interpolation based on %s for cluster %s: %w", property, cluster.Name, err)
+	}
+	if q == nil {
+		// The property is not available for the cluster.
+		return 0, nil
+	}
+
+	// Read the pre-prepared min/max values from the state, calculated in the PreScore stage.
+	mm, ok := state.minMaxValuesByProperty[property]
+	if !ok {
+		return 0, fmt.Errorf("failed to look up extremums for property %s, no state is prepared", property)
+	}
+	if mm.min == nil || mm.max == nil {
+		// The extremums are not available; this can happen when none of the clusters support
+		// the property.
+		//
+		// Normally this will never occur as the check before has guaranteed that at least
+		// observation has been made.
+		return 0, fmt.Errorf("extremums for property %s are not available, yet a reading can be found from cluster %s", property, cluster.Name)
+	}
+	minQ, maxQ := mm.min, mm.max
+
+	// Cast the quantities as floats to allow ratio estimation.
+	//
+	// This conversion will incur precision loss, though in most cases such loss has very limited
+	// impact.
+	f := q.AsApproximateFloat64()
```

Comment:
nit: what does f stand for?

Reply:
Oh, it's just a float value. Would you like me to use a different name?